### PR TITLE
AM2R: Add option to place DNA anywhere

### PR DESCRIPTION
**Pull request by JeffGainsNGames**
Allows DNA to be placed in any location (including off World).

### CODE REVIEW

**Comment by JeffGainsNGames** [comment]
Need to look in to:
- running into an issue with radio buttons not behaving nicely
- generator fails often with higher DNA, doesn't seem to happen when only on metroids/bosses

File: randovania/games/am2r/gui/preset_settings/am2r_goal_tab.py [hunk 1]
```diff
@@ -50,12 +52,38 @@ def _edit_config(self, call: Callable[[AM2RArtifactConfig], AM2RArtifactConfig])
...
     def num_preferred_locations(self) -> int:
+        if self.free_placement_radiobutton.isChecked():
+            return 46
         if self.prefer_metroids_check.isChecked():
             return 46
         if self.prefer_bosses_check.isChecked():
             return 6
         return 0
 
+    def _on_restrict_placement(self, value: bool) -> None:
```
[the remark this answers is not in the document]
Comment (JeffGainsNGames) [reply randovania/games/am2r/gui/preset_settings/am2r_goal_tab.py:63]: is this a new test that goes in `/test_preset_am2r_goal.py`? Or does this get built in to the existing test?

File: randovania/games/am2r/layout/preset_describer.py [hunk 2]
```diff
@@ -19,7 +19,16 @@
 
 def describe_artifacts(artifacts: AM2RArtifactConfig) -> list[dict[str, bool]]:
     has_artifacts = artifacts.required_artifacts > 0
-    if has_artifacts:
+    if has_artifacts and artifacts.prefer_anywhere:
+        return [
```
[the remark this answers is not in the document]
Comment (JeffGainsNGames) [reply randovania/games/am2r/layout/preset_describer.py:23]: is this a new test that goes in `/test_am2r_describer.py`? Or does this get built in to the existing test?

File: randovania/games/am2r/generator/bootstrap.py [hunk 3]
```diff
@@ -28,7 +28,9 @@ def all_dna_locations(game: GameDescription, config: AM2RArtifactConfig):
         if isinstance(node, PickupNode):
             # Metroid pickups
             name = node.extra["object_name"]
-            if config.prefer_metroids and name.startswith("oItemDNA_"):
+            if config.prefer_anywhere:
+                locations.append(node)
```
[the remark this answers is not in the document]
Comment (JeffGainsNGames) [reply randovania/games/am2r/generator/bootstrap.py:32]: doesn't this set the list of possible nodes to use for DNA? if this was removed then how does DNA anywhere know where to place them? (unless I'm misunderstanding)

[the remark this answers is not in the document]
Comment (Miepee) [reply randovania/games/am2r/generator/bootstrap.py:32]: This snippet is part of the function `all_dna_locations`
And the function `all_dna_locations` is only ever called if DNA is *not* everywhere. see here:
https://github.com/randovania/randovania/pull/5565/files/fa79f1eb393e30ef31d207adbba900fe88947ed9#diff-e5ccc83827246fd6cbc9ec3e09472b2e9b73eaae0366402005035316704f2ec4R78-R86

File: randovania/games/am2r/layout/preset_describer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Miepee) [reply randovania/games/am2r/layout/preset_describer.py:23]: you can probably make it go into the existing test.

File: randovania/games/am2r/gui/preset_settings/am2r_goal_tab.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Miepee) [reply randovania/games/am2r/gui/preset_settings/am2r_goal_tab.py:63]: yeah you want new tests in there

File: test/games/am2r/layout/test_am2r_describer.py [hunk 4]
```diff
@@ -12,23 +12,38 @@
...
 
-@pytest.mark.parametrize(("has_artifacts"), [(False), (True)])
-def test_am2r_format_params(has_artifacts: bool):
+@pytest.mark.parametrize(
+    ("artifacts"),
+    [
+        AM2RArtifactConfig(True, True, False, 5),
+        AM2RArtifactConfig(True, False, False, 46),
+        AM2RArtifactConfig(False, True, False, 6),
+        AM2RArtifactConfig(False, False, False, 0),
```
Comment (Miepee) [inline test/games/am2r/layout/test_am2r_describer.py:21]: you should add some more test cases with `prefer_anywhere` being set to True.

File: test/games/am2r/gui/preset_settings/test_preset_am2r_goal.py [hunk 5]
```diff
@@ -65,3 +65,39 @@ def test_preferred_dna(
...
+    tab.on_preset_changed(preset)
+
+    tab.free_placement_radiobutton.setChecked(True)
+
+    # Run
+    tab.restrict_placement_radiobutton.setChecked(True)
+
+    # Assert
+    assert tab.prefer_metroids_check.isEnabled()
+    assert tab.prefer_bosses_check.isEnabled()
```
[the remark this answers is not in the document]
Comment (JeffGainsNGames) [reply test/games/am2r/gui/preset_settings/test_preset_am2r_goal.py:102]: My understanding was that this was confirming that the buttons were selectable and therefore the `free_placement` radio button worked

File: test/games/am2r/gui/preset_settings/test_preset_am2r_goal.py [hunk 6]
```diff
@@ -65,3 +65,39 @@ def test_preferred_dna(
 
     assert editor.configuration.artifacts.required_artifacts == expected_artifacts
     assert tab.dna_slider.value() == expected_artifacts
+
+
+@pytest.mark.parametrize(
+    ("prefer_metroids", "prefer_bosses", "prefer_anywhere", "expected_artifacts"),
```
[the remark this answers is not in the document]
Comment (JeffGainsNGames) [reply test/games/am2r/gui/preset_settings/test_preset_am2r_goal.py:71]: no, but I could not figure out how to run a test without parameters (unless I'm dumb and just remove this line)

[the remark this answers is not in the document]
Comment (Miepee) [reply test/games/am2r/gui/preset_settings/test_preset_am2r_goal.py:71]: you basically remove all of these lines:
```
@pytest.mark.parametrize(
    ("prefer_metroids", "prefer_bosses", "prefer_anywhere", "expected_artifacts"),
        [(True, False, False, 10)],
)
```

and then you also remove them from the functions signature, like so:
```
def test_restricted_placement(
    skip_qtbot,
    am2r_game_description,
    preset_manager,
):
```

File: test/games/am2r/gui/preset_settings/test_preset_am2r_goal.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Miepee) [reply test/games/am2r/gui/preset_settings/test_preset_am2r_goal.py:102]: oh, enabled! i read checked.
You should then also assert that the restrict placement radio button is checked.

File: test/games/am2r/gui/preset_settings/test_preset_am2r_goal.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Miepee) [reply test/games/am2r/gui/preset_settings/test_preset_am2r_goal.py:71]: if you remove the expected artifacts from the parameters, you change the editor.configuration.required_artifacts assert to be the same as it was before clicking the radio button

[the remark this answers is not in the document]
Comment (JeffGainsNGames) [reply test/games/am2r/gui/preset_settings/test_preset_am2r_goal.py:71]: Do I even really need that assert?

[the remark this answers is not in the document]
Comment (Miepee) [reply test/games/am2r/gui/preset_settings/test_preset_am2r_goal.py:71]: you should. we don't want clicking radio buttons to suddenly change the artifact count. so we assert for it.

File: test/games/am2r/gui/preset_settings/test_preset_am2r_goal.py [hunk 7]
```diff
@@ -65,3 +65,63 @@ def test_preferred_dna(
...
+    base = preset_manager.default_preset_for_game(game).get_preset()
+    preset = dataclasses.replace(base, uuid=uuid.UUID("b41fde84-1f57-4b79-8cd6-3e5a78077fa6"))
+    base_configuration = preset.configuration
+    options = MagicMock()
+    assert isinstance(base_configuration, AM2RConfiguration)
+
+    tab = PresetAM2RGoal(editor := PresetEditor(preset, options), am2r_game_description, MagicMock())
+    skip_qtbot.addWidget(tab)
+    tab.on_preset_changed(preset)
+    artifact_count = editor.configuration.artifacts.required_artifacts
```
Comment (Miepee) [inline test/games/am2r/gui/preset_settings/test_preset_am2r_goal.py:86]: mypy error.
You might want to move the assert from below somewhere further to the top.

File: test/games/am2r/gui/preset_settings/test_preset_am2r_goal.py [hunk 8]
```diff
@@ -65,3 +65,63 @@ def test_preferred_dna(
...
```
Comment (Miepee) [inline test/games/am2r/gui/preset_settings/test_preset_am2r_goal.py:116]: see other mypy error.

File: test/test_files/log_files/am2r/door_lock.rdvgame
Comment (Miepee) [inline test/test_files/log_files/am2r/door_lock.rdvgame:1]: can you please redo these changes for the test files here? the preset migration should take care of it.

**Review by Miepee — APPROVED** [review]
Thank you for pushing through with this!

**Pull request merged**